capellambse/_diagram_cache.py

```diff
@@ -219,6 +220,255 @@ def _sanitize_filename(fname: str) -> str:
+        (x_min, x_max, y_min, y_max)
+    """
+    return polyline_extents(element)
+
+
+def polyline_extents(
+    element: etree.Element,
+) -> tuple[float, float, float, float]:
+    """Compute extents for a polyline.
+
+    Parameters
+    ----------
+    element : etree.Element
+        The XML element representing the polyline.
+
+    Returns
+    -------
+    tuple[float, float, float, float]
+        (x_min, x_max, y_min, y_max)
+    """
+    points = element.get("points", "")
+    points = [tuple(map(float, p.split(","))) for p in points.strip().split()]
+    x_coords, y_coords = zip(*points)
+    return (min(x_coords), max(x_coords), min(y_coords), max(y_coords))
+
+
+def rect_extents(element: etree.Element) -> tuple[float, float, float, float]:
+    """Compute extents for a rectangle.
+
+    Parameters
+    ----------
+    element : etree.Element
+        The XML element representing the rectangle.
+
+    Returns
+    -------
+    tuple[float, float, float, float]
+        (x_min, x_max, y_min, y_max)
+    """
+    width_str = element.get("width", "0")
+    height_str = element.get("height", "0")
+    try:
+        width = float(width_str)
+        height = float(height_str)
+    except ValueError:
+        return -1.0, -1.0, -1.0, -1.0
+    x = float(element.get("x", 0))
+    y = float(element.get("y", 0))
+    width = float(element.get("width", 0))
+    height = float(element.get("height", 0))
+    return (x, x + width, y, y + height)
+
+
+def text_extents(element: etree.Element) -> tuple[float, float, float, float]:
+    """Compute extents for text.
+
+    Simplified assumption of width based on character count and font
+    size.
+
+    Parameters
+    ----------
+    element : etree.Element
+        The XML element representing the text.
+
+    Returns
+    -------
+    tuple[float, float, float, float]
+        (x_min, x_max, y_min, y_max)
+    """
+    x = float(element.get("x", 0))
+    y = float(element.get("y", 0))
+    text = element.text or ""
+    font_size = float(element.get("font-size", 16))  # Default font size
+    width = len(text) * font_size * 0.6  # Simplistic width estimation
+    return (x, x + width, y - font_size, y)
+
+
+def _svg_content_bbox(
+    root: lxml.etree.Element,
+) -> tuple[float, float, float, float]:
+    x_min = y_min = float("inf")
+    x_max = y_max = float("-inf")
+    funcs = set()
+    for element in root.iter():
+        # pylint: disable-next=c-extension-no-member,protected-access
+        if isinstance(element, lxml.etree._Comment):
+            continue
+        if not hasattr(element, "tag"):
+            continue
+        # pylint: disable-next=c-extension-no-member
+        qname = lxml.etree.QName(element.tag)
+        if not hasattr(qname, "localname"):
+            continue
+        func_name = f"{qname.localname}_extents"
+        if func_name not in globals():
+            continue
+        funcs.add(func_name)
+        func = globals()[func_name]
```

Comment:
Instead of having a check and the function retrieval from `globals()` we can combine this into one:
```suggestion
        logger.debug("Request %r from SVG shape extenders", func_name)
        if not (func := globals().get(func_name)):
            logger.error("Unknown SVG shape extender %r requested", func_name)
            continue
```
Side-effect: No `KeyError` for unknown func_name raised anymore!

Also globals is used here for the function registry which is inefficient but I wouldn't call it an anti-pattern yet. globals will just grow and access from it will become worse. We have 8 ´*_extents` objects and without the imports 15 other functions in globals. Declaring a specific mapping as registry for these SVG shape extenders and retrieving from it is a good refactoring. It also serves as a good overview of covered shapes. :)

Reply:
unknown extender is not an error. it is a known limitation. thanks for the `:=`. will do that!

Reply:
The main benefit of having a dedicated lookup dict IMO is actually the ability to statically type it (did I mention I like static typing?) - for example:

```python
def _circle_extents(elem: etree._Element) -> Extents: ...
def _polygon_extents(elem: etree._Element) -> Extents: ...
# etc.

EXTENT_FUNCTIONS: dict[str, t.Callable[[etree._Element], Extents]] = {
    "circle": _circle_extents,
    "polygon": _polygon_extents,
    "polyline": _polygon_extents,
    # ...
}

def some_func():
    for elem in root.iter:
        reveal_type(globals()[f"_{elem.tag}_extents"](elem))  # Any
        reveal_type(EXTENT_FUNCTIONS[elem.tag](elem))  # Extents
```

Unlike `globals()` lookups, which always become `Any`, this allows mypy to fully understand the entire call chain and therefore provide actual static analysis results - especially if at any point the return value of the functions should change.